google/appengine/api/urlfetch/UrlFetchStream.php

```
@@ -0,0 +1,429 @@
+     * @param int $options_stream: UNUSED in the context of URLs.
+     * @param string $opened_path: UNUSED in the context of URLs.
+     *
+     * @throws \Exception if URLFetch request is nto successful.
+     *
+     * @return bool Returns true on success or false on failure.
+     *
+     */
+    public function stream_open($url, $mode, $options_stream, &$opened_path)
+    {
+        if ($this->context == null) {
+            throw new Exception('ERROR: Context not set for stream wrapper.');
+        }
+        $options = stream_context_get_options($this->context);
+
+        foreach ($options as $web_protocol => $context_array) {
+            foreach ($context_array as $context_key => $context_value) {
+                $this->setContextOptions($context_key, $context_value);
+            }
+        }
+
+        // Check and store User-agent if specified separately from header.
+        if (!in_array('User-Agent', $this->headers) && $this->user_agent != '') {
+            $header_arr = ['User-Agent' => $this->user_agent];
+            $this->headers = $this->headers + $header_arr;
+        }
+
+        try {
+            $urlfetch = new UrlFetch();
+            $resp =
+                $urlfetch->fetch(
+                    $url,
+                    $this->method,
+                    $this->headers,
+                    $this->content,
+                    true,
+                    true,
+                    $this->timeout,
+                    $this->verify_peer);
+            $this->url_fetch_response = $resp;
+            $this->stream = new CachingStream(Stream::factory($resp->getContent()));
+            $this->response_headers = $this->buildHeaderArray($resp->getStatuscode(), $resp->getHeaderList());
+        } catch (Exception $e) {
+            echo '<br>Caught exception:<br>';
+            var_dump($e->getMessage());
+            echo '<br><br>';
+            var_dump($e->getTrace());
+            exit();
```

Don't call `exit()` here. The right thing to do is return `false` on error, and if you want you can also surface the error using `trigger_error`. For  an example see the GCS stream wrapper:

- https://github.com/googleapis/google-cloud-php/blob/2fc2caf8ea1eed6b0ce2f65d1b664b9e54643ef1/Storage/src/StreamWrapper.php#L241
- https://github.com/googleapis/google-cloud-php/blob/2fc2caf8ea1eed6b0ce2f65d1b664b9e54643ef1/Storage/src/StreamWrapper.php#L804:22

Also don't print anything here.

---

Done